In [45]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.model_selection import KFold,train_test_split,cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

In [28]:
data = pd.read_csv(r"diabetes.csv")

In [29]:
#EDA

In [30]:
data.columns

Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')

In [31]:
#data.drop(['PassengerId','Name','Ticket','Cabin'],axis=1,inplace=True)

In [32]:
#data = data.ffill()

In [33]:
data.dtypes

Pregnancies                   int64
Glucose                       int64
BloodPressure                 int64
SkinThickness                 int64
Insulin                       int64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                       int64
dtype: object

In [34]:
#le = LabelEncoder()
#data[['Sex','Embarked']] = data[['Sex','Embarked']].apply(le.fit_transform)

In [35]:
x = data.drop('Outcome',axis=1)
X = StandardScaler().fit_transform(x)
y = data.Outcome

In [36]:
kf = KFold(n_splits=20)     #5 fold cross validation

In [37]:
s_model = SVC()

In [38]:
D_model = DecisionTreeClassifier()

In [39]:
R_model = RandomForestClassifier(n_estimators=500)

In [40]:
#5 fold validation of svm

In [41]:
li_s = []
for train_index,test_index in kf.split(x,y):
    xtrain,xtest,ytrain,ytest = x.loc[train_index],x.loc[test_index],y.loc[train_index],y.loc[test_index]
    s_model.fit(xtrain,ytrain)
    li_s.append(s_model.score(xtest,ytest))
    
print(li_s)
print("mean: ",sum(li_s)/len(li_s))


[0.6923076923076923, 0.717948717948718, 0.7948717948717948, 0.7948717948717948, 0.7435897435897436, 0.6666666666666666, 0.7435897435897436, 0.6153846153846154, 0.6842105263157895, 0.8421052631578947, 0.8157894736842105, 0.868421052631579, 0.6842105263157895, 0.8947368421052632, 0.7631578947368421, 0.9210526315789473, 0.7631578947368421, 0.7631578947368421, 0.7894736842105263, 0.7368421052631579]
mean:  0.7647773279352226


In [42]:
li_D = []
for train_index,test_index in kf.split(x,y):
    x_train,x_test,y_train,y_test = x.loc[train_index],x.loc[test_index],y.loc[train_index],y.loc[test_index]
    D_model.fit(x_train,y_train)
    li_D.append(D_model.score(x_test,y_test))
    
print(li_D)
print("mean : ",sum(li_D)/len(li_D))

[0.6410256410256411, 0.5897435897435898, 0.7435897435897436, 0.6666666666666666, 0.6923076923076923, 0.5641025641025641, 0.6153846153846154, 0.5641025641025641, 0.7105263157894737, 0.7105263157894737, 0.7368421052631579, 0.7368421052631579, 0.7631578947368421, 0.7894736842105263, 0.7631578947368421, 0.7631578947368421, 0.7631578947368421, 0.5526315789473685, 0.7368421052631579, 0.6578947368421053]
mean :  0.6880566801619432


In [43]:
li_s = []
for train_index,test_index in kf.split(x,y):
    xtrain,xtest,ytrain,ytest = x.loc[train_index],x.loc[test_index],y.loc[train_index],y.loc[test_index]
    R_model.fit(xtrain,ytrain)
    li_s.append(R_model.score(xtest,ytest))
    
print(li_s)
print("mean: ",sum(li_s)/len(li_s))

[0.6923076923076923, 0.6923076923076923, 0.8461538461538461, 0.7948717948717948, 0.7692307692307693, 0.6666666666666666, 0.6666666666666666, 0.5897435897435898, 0.7105263157894737, 0.8421052631578947, 0.868421052631579, 0.8157894736842105, 0.7105263157894737, 0.9210526315789473, 0.7894736842105263, 0.9210526315789473, 0.7631578947368421, 0.6842105263157895, 0.7894736842105263, 0.7894736842105263]
mean:  0.7661605937921727


In [44]:
models = [s_model,D_model,R_model]
names = ['svm','DecisionTree','Randomforest']
for i in range(len(models)):
    li = []
    for train_index,test_index in kf.split(x,y):
        x_train,x_test,y_train,y_test = x.loc[train_index],x.loc[test_index],y.loc[train_index],y.loc[test_index]
        models[i].fit(x_train,y_train)
        li.append(models[i].score(x_test,y_test))
        
    print(li)
    #print("mean : ",names[i],sum(li)/len(li))
    print("mean : ",names[i],max(li))

[0.6923076923076923, 0.717948717948718, 0.7948717948717948, 0.7948717948717948, 0.7435897435897436, 0.6666666666666666, 0.7435897435897436, 0.6153846153846154, 0.6842105263157895, 0.8421052631578947, 0.8157894736842105, 0.868421052631579, 0.6842105263157895, 0.8947368421052632, 0.7631578947368421, 0.9210526315789473, 0.7631578947368421, 0.7631578947368421, 0.7894736842105263, 0.7368421052631579]
mean :  svm 0.9210526315789473
[0.6666666666666666, 0.6666666666666666, 0.7948717948717948, 0.6666666666666666, 0.6923076923076923, 0.5128205128205128, 0.6410256410256411, 0.5641025641025641, 0.7105263157894737, 0.7105263157894737, 0.7368421052631579, 0.7631578947368421, 0.7105263157894737, 0.7631578947368421, 0.6842105263157895, 0.7894736842105263, 0.7631578947368421, 0.6052631578947368, 0.7631578947368421, 0.6578947368421053]
mean :  DecisionTree 0.7948717948717948
[0.6923076923076923, 0.6923076923076923, 0.8717948717948718, 0.7948717948717948, 0.7692307692307693, 0.6923076923076923, 0.641025

Random forest gives 92% accuracy for 12 fold cross validation.